Windows: Add ETW logging driver plug-in

cednation opened the pull request:
Signed-off-by: Cedric Davies cedricda@microsoft.com
@jhowardmsft 

Adds an ETW logging driver to Docker on Windows. ETW stands for Event Tracing in Windows, and is the common framework for tracing Windows applications. This log driver forwards each container log message as an ETW event. A client can then create an ETW listener to listen for these events.

A client registers to listen to events from our ETW provider, by specifying our provider's GUID, which is: a3693192-9ed6-46d2-a981-f8226c8363bd. Here is an example of how to listen to these events using the logman utility available in any Windows installation:
1. logman start -ets DockerContainerLogs -p {a3693192-9ed6-46d2-a981-f8226c8363bd} 0 0 -o trace.etl
2. Run your container(s) with the etwlogs driver, by adding --log-driver=etwlogs to the Docker run command, and generate log messages
3. logman stop -ets DockerContainerLogs
4. You can then convert the generated etl file to XML using: tracerpt -y trace.etl

Each ETW event generated will contain a data string in this format:
container_name: %s, image_name: %s, container_id: %s, image_id: %s, source: [stdout | stderr], log: %s

Your ETW listener can then parse this string, to both get the log message generated, as well as its context information, such as which container emitted the log message. The Timestamp is also available within the ETW event.

cednation committed:
Windows: Add ETW logging driver plug-in

Signed-off-by: Cedric Davies <cedricda@microsoft.com>

MHBauer commented:
You might as well squash the three commits together. Having golint and gofmt in separate commits doesn't really add anything helpful to the history. Thanks for the contribution!

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "etwlogs" git@github.com:cednation/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

cednation commented:
@MHBauer Squashed the commits into one. Thanks!

calavera commented:
:raised_hands: LGTM :raised_hands:

MHBauer commented:
@cednation Looking good. Thanks for the contribution. What will really polish this off is to add an entry to `docs/admin/logging/overview.md` and perhaps a docs page of it's own.

I wish I knew more about windows programming, but I have heard the praises of ETW being sung. I can't give much more than general thoughts about the code itself. It looks complete for an implementation of a logger. I like that there are debug logs.

I feel like I have seen the `fixContainerName` function elsewhere, and wonder if there is a utility that can be provided, or a potential for a new function on the Container object (That change is out of scope here, but that's my thought).

MHBauer commented on daemon/logger/etwlogs/etwlogs_windows.go:
```diff
@@ -0,0 +1,183 @@
+		return err
+	}
+	// The provider's GUID is {a3693192-9ed6-46d2-a981-f8226c8363bd}
+	guid := syscall.GUID{
```
Is `syscall.GUID` only in the Windows version of golang? I had problems trying to find documentation.

cednation replied:
@MHBauer Hi MGBauer, thanks for the review comments! Your suggestion to add documentation to the overview.md file is a great one. I've done that, and also added more detailed documentation in a new etwlogs.md file (trying to follow the format used by the other logging drivers). I've updated the pull request with these additional changes. 

As per syscall.GUID, yes I believe this is only in the Windows version. I see the type is defined within syscall/ztypes_windows.go. It is common to use GUIDs within Windows programming.

MHBauer commented on docs/reference/logging/etwlogs.md:
```diff
@@ -0,0 +1,74 @@
+an ETW listener to listen to these events. 
+
+The ETW provider that this logging driver registers with Windows, has the 
+GUID identifier of: a3693192-9ed6-46d2-a981-f8226c8363bd. A client creates an 
```
I would say verbatim the GUID with backtics.

MHBauer commented on docs/reference/logging/etwlogs.md:
```diff
@@ -0,0 +1,74 @@
+
+| Field                | Description                               |
+--------------------------------------------------------------------
+| container_name       | The container name at the time it was 
```
This table isn't rendering right. I'm not sure what it needs to fix it.

MHBauer commented on docs/reference/logging/etwlogs.md:
```diff
@@ -0,0 +1,74 @@
+| log                  | The container's log message.              |
+
+Here is an example string:
+container_name: backstabbing_spence, 
```
If you want this to be a code block, there needs to be a newline and then indent four spaces.

MHBauer commented on docs/reference/logging/etwlogs.md:
```diff
@@ -0,0 +1,74 @@
+The following regular expression may be helpful to you for parsing and extracting the data 
+from the message (you may need to modify it slightly to fit your programming 
+language of choice):
+"container_name: ([\w_\d]+), image_name: ([\w_\d]+), container_id: ([\da-f]+), image_id: (.*:)?([\da-f]+), source: (\w+), log: (.*)"
```
same, newline, indent four spaces.

MHBauer commented:
@cednation good start on the docs. :+1:  I've left a couple of comments. To check how the documentation renders, issue a `make docs` and then open a browser to the address it spits out at the end.

cednation commented:
@MHBauer Thanks for the helpful comments and your patience. I never modified a md file before and wasn't sure how to test the rendering. I've fixed and updated everything you recommended and tested the rendering. I fixed the table, and made other items verbatim strings or code snippets as appropriate. And replaced all my apostrophes with backticks. :) Please let me know if you see any other things in the docs that need updating. 

Oh and on the utility method for fixContainerName, I'm happy to switch to a common utility method if there is one, but I couldn't find one. I agree there probably should be, but I don't know if it would be appropriate for me to add it. 
I have noticed that at least one other logging driver strips out the leading '/' as I am doing, but some others, for example fluentd, do not. And of course the docker client strips it out when you do a `dockder ps`. Personally, I feel the leading '/' is not helpful for the user to see, but I wouldn't mind keeping it in the output if you feel that is more consistent, I would just update my listener's regex to strip it out.

MHBauer commented on docs/reference/logging/etwlogs.md:
```diff
@@ -0,0 +1,77 @@
+
+The ETW provider that this logging driver registers with Windows, has the 
+GUID identifier of: `{a3693192-9ed6-46d2-a981-f8226c8363bd}`. A client creates an 
+ETW listener and registers to listen to events from the logging driver`s provider. 
```
this backtick ` can be a regular single quote '

MHBauer commented:
I think the docs are looking real good, let's get @thaJeztah & @moxiegirl to take a look.

I wouldn't worry about the container name fix now. I'll put it in my list of things to follow up on. This is a very nicely single self-contained piece of work. Let's keep it like that for now.

Thanks for being so understanding and accommodating of changes.

thaJeztah commented:
@MHBauer yup, docs already look good (at a glance); but let's get it moved to "docs review" first

@cednation looks like this needs a rebase :cry:

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

cednation commented:
@thaJeztah @MHBauer I'm very confused on how to properly do a rebase. Can you offer any advice? I don't understand why when I updated the docs last time, did a commit --amend, and then 'git push -f' that that didn't work. (I know you would want me to amend the older commit, so that there is only one commit in the history)
I tried to fix things by doing a git fetch upstream, git pull, and then git rebase -i. But since there were about 200 entries here, I thought I could do the rebase faster non-interactively, and did 'git rebase'. Now when I push, I get 98 commits being pushed. 
I'm sorry, Git is very new to me as you can probably tell. 
Do you know what I should do now to fix things? Or should I just close this review and start all over from scratch?

cednation commented:
@thaJeztah @MHBauer Also, when I was trying to do the interactive rebase, it said I couldn't squash the first commit, which wasn't mine, but I thought I was supposed to squash every commit other than mine.... so wasn't sure how to proceed.

MHBauer commented:
This looks like you rebased master on top of your branch instead of your branch onto master. Rebases are directional. I usually:
```
git checkout ${MY_BRANCH}
git fetch
git rebase origin/master
```
I'm going to need to think a minute about your case.

thaJeztah commented:
@cednation you can find some instructions in our contributors guide; https://docs.docker.com/opensource/workflow/work-issue/#pull-and-rebase-frequently

cednation commented:
@thaJeztah Those are the instructions that got me screwed up in the first place! :) 
I followed those instructions, but when I got to "Replace the pick keyword with squash on all but the first commit", I got confused about following that, because the first commit wasn't my commit. Also there were over 100 commits to squash, so I thought there must be an easier way. (On Monday when I did the same thing with an older branch, there were over 600 commits, after doing a search & replace, it still didn't work right, and I was forced to start over.)
So then I tried just "git rebase" and that's what ended up getting me into the state that I am in. Not sure how to reverse course now... Maybe I need to delete everything and start over?

MHBauer commented:
Alright, be very very careful. If it breaks, you get to keep both pieces!
```
git reset --hard 56e5a26251b0fe3ea51532a31527af22967c8fba   # This is your commit
git push -f
```
Make sure to save any work you have in progress in your local copy. Zip up the directory ahead of time if you want.

I did that and went from https://github.com/MHBauer/docker/commits/etw-docs to https://github.com/MHBauer/docker/commits/etw-docs-rebased

MHBauer commented:
I forgot to rebase in that, sorry! You can contact me if a chat would help! Sorry!

jhowardmsft commented:
@cednation I sent you some instructions which will hopefully help. Otherwise, catch me on Monday and I'll walk you through it :)

cednation commented:
@MHBauer @jhowardmsft Thanks for your help!! I used the combined information that you sent me to fix my git history and resubmit.

cednation commented:
@MHBauer @thaJeztah One question I have when you are reviewing the etwlogs.md file, do you think it is worthwhile for me to include the regular expression help in the document, even though maybe it isn't the *best* regex expression that could be used. For instance, I originally had a mistake in it, and that I didn't realize '.' and '-' were valid characters for container names, which I have since updated. Maybe I should just remove it altogether, or simplify it to capture everything up to the comma, in case the allowable characters in a container name ever changes.

MHBauer commented:
@cednation Given it's a regex for the message above. Maybe not explicitly, but maybe mention the possibility? I don't have enough experience with consuming the output of etw logs to know if this is a standard thing. It seems to be an outlier in that it is just one large string vs a message with explicitly defined fields that arrive at the receiver. You do have the `Note` that follows, which explains to me *why* I would use a regex.

cednation commented:
@MHBauer As I think about the regex expression in the doc more, I think it is not that helpful to have it. There are many different ways to parse a string, and different ways to construct the regex expression you use, plus different formats for regex expressions in different languages. This just happens to be the one I'm using right now in my listener app, but I may change it later. It's something that may work in a blog post or something like that, but I think I'll remove it from the the official documentation .md file.

MHBauer commented:
LGTM. Good Luck! Happy Hacking!

calavera commented:
There are already two LGTM from maintainers for this change. Moved to docs review.

calavera commented:
@thaJeztah, @vdemeester can you take a look at the docs?

vdemeester commented:
@cednation You need to update a little bit more on documentation (reusing @thaJeztah comment :sweat_smile:)

- A link on the index page; https://github.com/docker/docker/blob/master/docs/admin/logging/index.md

Also, not sure if this driver is usable for `docker logs`; if it is, it needs a mention in https://github.com/docker/docker/blob/master/docs/reference/commandline/logs.md and https://github.com/docker/docker/blob/master/man/docker-logs.1.md

Finally, the driver needs to be added as option for `--log-driver` in the man-pages:

- https://github.com/docker/docker/blob/master/man/docker-run.1.md
- https://github.com/docker/docker/blob/master/man/docker-create.1.md
- https://github.com/docker/docker/blob/master/man/docker-daemon.8.md

And the command reference:

- https://github.com/docker/docker/blob/master/docs/reference/commandline/daemon.md
- https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md
- https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md

And the API docs :
- https://github.com/docker/docker/blame/master/docs/reference/api/docker_remote_api_v1.23.md#L402

There might be need to add/update completion for powershell (but I'm not sure) and updating the completion for ˙zsh˙ and `bash` depending on the host (but not sure if we do that or not @albers @sdurrheimer)

albers commented:
If you want, I can take over the bash completion part. Just give me a ping.

cednation commented:
@vdemeester Thanks, I will work on making those changes. 

@albers Yes, if you could take over the bash completion part, or just give me instructions on how to do it, that would be great, as I'm not sure how to do it.

cednation commented:
@vdemeester I made the suggested updates to the documentation. 2 things to note:
1. I didn't see anywhere to update in 3 of the files you mentioned in the command reference:
https://github.com/docker/docker/blob/master/docs/reference/commandline/daemon.md
https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md
https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md
Please let me know if I just missed it. I did a search on other log drivers and didn't find anything for them. 

2. In file https://github.com/docker/docker/blame/master/docs/reference/api/docker_remote_api_v1.23.md#L402, I noticed that fluentd was missing from the list of logging driver options, so I added it. 

@albers Are you still working on the bash completion part?

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1024/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.2 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.6%
 18.3%
 22.0%
 25.7%
 29.3%
 33.0%
 36.7%
 40.4%
 44.0%
 47.7%
 51.4%
 55.1%
 58.7%
 62.4%
 66.1%
 69.8%
 73.4%
 77.1%
 80.8%
 84.5%
 88.1%
 91.8%
 95.5%
 99.2%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.42 m   54.42 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1164685795 Bytes/sec.
   Speed :            66643.855 MegaBytes/min.
   Ended : Sunday, February 14, 2016 11:59:29 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-9441
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1026/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    Newer     		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    Newer     		  27.2 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.6%
 18.3%
 22.0%
 25.7%
 29.3%
 33.0%
 36.7%
 40.4%
 44.0%
 47.7%
 51.4%
 55.1%
 58.7%
 62.4%
 66.1%
 69.8%
 73.4%
 77.1%
 80.8%
 84.5%
 88.1%
 91.8%
 95.5%
 99.2%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         1
   Bytes :   54.42 m   54.42 m         0         0         0   27.21 m
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1056844518 Bytes/sec.
   Speed :            60473.128 MegaBytes/min.
   Ended : Monday, February 15, 2016 7:35:21 AM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-94413
---

~~~

cednation commented:
The windowsTP4 failure doesn't look like something I would have caused. Problem with the build machine?

ln: failed to create hard link '/d/CI/CI-94413db/binary/docker-94413db.exe': File exists
07:35:23 ERROR: Failed to link

albers commented:
@cednation I'll add the bash completion after this PR is merged in order to avoid unnecessary dependencies among PRs.

thaJeztah commented:
Restarting WindowsTP4 (https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1026/console)

thaJeztah commented on docs/admin/logging/etwlogs.md:
```diff
@@ -0,0 +1,69 @@
+
+The ETW provider that this logging driver registers with Windows, has the 
+GUID identifier of: '{a3693192-9ed6-46d2-a981-f8226c8363bd}'. A client creates an 
+ETW listener and registers to listen to events from the logging driver`s provider. 
```
Can you change the back-tick in `driver's` to a regular quote?

thaJeztah commented on docs/admin/logging/etwlogs.md:
```diff
@@ -0,0 +1,69 @@
+an ETW listener to listen to these events. 
+
+The ETW provider that this logging driver registers with Windows, has the 
+GUID identifier of: '{a3693192-9ed6-46d2-a981-f8226c8363bd}'. A client creates an 
```
I think we should put the GUID here in back-ticks to make it stand out more;

`{a3693192-9ed6-46d2-a981-f8226c8363bd}`

thaJeztah commented on docs/admin/logging/etwlogs.md:
```diff
@@ -0,0 +1,69 @@
+| Field                | Description                                     |
+-----------------------|-------------------------------------------------|
+| `container_name`     | The container name at the time it was started.  |
+| `image_name`         | The name of the container`s image.              |
```
same here (back-tick in `container's`)

thaJeztah commented on docs/admin/logging/etwlogs.md:
```diff
@@ -0,0 +1,69 @@
+| `container_name`     | The container name at the time it was started.  |
+| `image_name`         | The name of the container`s image.              |
+| `container_id`       | The full 64-character container ID.             |
+| `image_id`           | The full ID of the container`s image.           |
```
And here (back-tick in `container's`)

thaJeztah commented on docs/admin/logging/etwlogs.md:
```diff
@@ -0,0 +1,69 @@
+
+Note that unlike many other ETW providers, because this provider emits only a message string, 
+and not a specially structured ETW event, it does not require a manifest file to be registered
+with the system in order to read and interpret its ETW events. 
```
This should be formatted in our standard "note" format. Perhaps write it like this;

```Markdown
> **Note**: this ETW provider emits only a message string, and not a specially
> structured ETW event. Therefore, it is not required to register a manifest file
> with the system to read and interpret the ETW events.
```

thaJeztah commented:
>  I didn't see anywhere to update in 3 of the files you mentioned in the command reference

That would be my mistake; I thought we also mentioned the available types there, just like in the man pages (perhaps we should)

thaJeztah commented:
Some nits, but docs overall looking good already

cednation commented:
@thaJeztah Thanks for all your helpful comments. I have made all of the changes that you suggested. Let me know if there is anything else that I should change.

thaJeztah commented:
docs LGTM (thanks @cednation)

ping @MHBauer @vdemeester PTAL

MHBauer commented on man/docker-daemon.8.md:
```diff
@@ -185,7 +185,7 @@ unix://[/path/to/socket] to use.
   Set key=value labels to the daemon (displayed in `docker info`)
 
-**--log-driver**="*json-file*|*syslog*|*journald*|*gelf*|*fluentd*|*awslogs*|*none*"
+**--log-driver**="*json-file*|*syslog*|*journald*|*gelf*|*fluentd*|*awslogs*|*etwlogs*|*none*"
```
Is there a reason `splunk` isn't in here and it is in the other two?

calavera replied:
I think someone forgot to add it here.

MHBauer commented:
I don't know enough about enabling logging on the daemon to know whether my nit is even a nit or not. LGTM otherwise.

cednation commented:
@MHBauer AFAIK the lists should be the same in all 3 cases. For the daemon it just specifies the default logging driver, so splunk should be in the list. I will add this minor change.

MHBauer commented:
re-LGTM. fingers crossed.

cednation commented:
@calavera Thanks!

thaJeztah commented:
@cednation congrats on your first PR merged! :tada:

MHBauer commented:
Welcome @cednation! Congratulations! :smiling_imp:
